core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/Utils.kt

```diff
@@ -337,28 +337,110 @@ internal fun <T : Comparable<T>> T.between(left: T, right: T, includeBoundaries:
         this > left && this < right
     }
 
-private const val DELIMITERS = "[_\\s]"
-public val DELIMITERS_REGEX: Regex = DELIMITERS.toRegex()
-public val DELIMITED_STRING_REGEX: Regex = ".+$DELIMITERS.+".toRegex()
+// Single regex to split words by non-alphanumeric characters, camelCase, and numbers
 
-internal val CAMEL_REGEX = "(?<=[a-zA-Z])[A-Z]".toRegex()
+internal val CAMEL_DEFAULT_DELIMITERS_REGEX =
+    "[^a-zA-Z0-9]+|(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])|(?<=\\d)(?=[a-zA-Z])|(?<=[a-zA-Z])(?=\\d)".toRegex()
 
-public fun String.toCamelCaseByDelimiters(delimiters: Regex): String = split(delimiters).joinToCamelCaseString()
+/**
+ * Converts a string into lowerCamelCase using [delimiters].
+ *
+ * - Splits words using a unified regex that handles delimiters.
+ * - If the string does not contain any letters or numbers, it remains unchanged.
+ * - Places underscore ("_") between numbers.
+ * - The first word remains in lowercase, and subsequent words are capitalized.
+ *
+ * Default behavior (with [CAMEL_DEFAULT_DELIMITERS_REGEX]):
+ *
+ * ```
+ * "hello_world" -> "helloWorld"
+ * "HelloWorld" -> "helloWorld"
+ * "json.parser.Config" -> "jsonParserConfig"
+ * "my.var_name test" -> "myVarNameTest"
+ * "thirdColumn" -> "thirdColumn"
+ * "someHTMLParser" -> "someHtmlParser"
+ * "RESTApi" -> "restApi"
+ * "OAuth2Token" -> "oAuth2Token"
+ * "GraphQLQuery" -> "graphQlQuery"
+ * "TCP_3_PROTOCOL" -> "tcp3Protocol"
+ * "123hello_world456" -> "123HelloWorld456"
+ * "API_Response_2023" -> "apiResponse2023"
+ * "UPPER_case-LOWER" -> "upperCaseLower"
+ * "12parse34CamelCase" -> "12Parse34CamelCase"
+ * "snake_case_example" -> "snakeCaseExample"
+ * "dot.separated.words" -> "dotSeparatedWords"
+ * "kebab-case-example" -> "kebabCaseExample"
+ * "MIXED_Case_with_123Numbers" -> "mixedCaseWith123Numbers"
+ * "___!!!___" -> "___!!!___"
+ * "1000.2000.3000" -> "1000_2000_3000"
+ * "UPPERCASE" -> "uppercase"
+ * "alreadyCamelCased" -> "alreadyCamelCased"
+ * "justNumbers123" -> "justNumbers123"
+ * "Just_Special\$Chars!!" -> "justSpecialChars"
+ * "singleword" -> "singleword"
+ * "word_with_underscores_and-dashes" -> "wordWithUnderscoresAndDashes"
+ * "10-20-aa" -> "10_20Aa"
+ * ```
+ *
+ * @return trhe formatted string in lowerCamelCase.
+ */
+public fun String.toCamelCaseByDelimiters(delimiters: Regex = CAMEL_DEFAULT_DELIMITERS_REGEX): String =
+    if (!this.any { it.isLetter() || it.isDigit() }) {
+        this // If the string has no letters, return it unchanged
+    } else {
+        split(delimiters)
+            .filter { it.isNotBlank() }
+            .map { it.lowercase() }
+            .joinNumbers("_")
+            .joinToCamelCaseString()
+    }
+
+/**
+ * Joins consecutive numbers in a list with the given [separator].
+ */
+private fun List<String>.joinNumbers(separator: CharSequence): List<String> {
```

Comment:
Hmm this seems a bit overcomplicated. It's not really clear what this function does and it's hard to check due to its complexity.
(What I think it does is [a1, 123, 345, bcd] -> [a1, 123_345, bcd], is that correct? This is then joined to "a1123_345Bcd". Actually, I think a1 and 123 should remain split as well.)

Wouldn't it be a lot simpler to perform a map-with-next and let `joinWithCamelCaseString` handle the joining part? Something like:
```kt
list.mapIndexed { i, current ->
    val next = list.getOrNull(i + 1) ?: return@mapIndexed current
    if (current.last().isDigit() && next.first().isDigit()) {
        current + "_"
    } else {
        current
    }
}
```
This would do [a1, 123, 345, bcd] -> [a1_, 123_, 345, bcd], after which they can be joined together like "a1_123_345Bcd"

Reply:
It assumes that all numbers and strings are separated, I'll write a comment.

Reply:
aah that was unclear indeed. Still, it might do with some simplification :)

Reply:
I'd prefer to separate digit logic